reckon-core/src/main/groovy/org/ajoberstar/reckon/core/GitVcs.java

```diff
@@ -0,0 +1,215 @@
+                    //     versions.add(new GitVersion(normal, mergeBase));
+                    // }
+                } else {
+                    logger.debug("Tag {} could not be parsed as a version.", tag.getName());
+                }
+            }
+
+            GitVersion any = findBase(walk, headCommit, versions.stream());
+            GitVersion normal = findBase(walk, headCommit, versions.stream().filter(ver -> ver.getVersion().getPreReleaseVersion().isEmpty()));
+            return new BaseVersion(normal.getVersion(), any.getVersion(), headCommit.equals(any.getRevCommit()));
+        } catch (IOException e) {
+            throw new UncheckedIOException(e);
+        }
+    }
+
+    private GitVersion findBase(RevWalk walk, RevCommit head, Stream<GitVersion> versions) {
+        try {
+            walk.reset();
+            walk.setRevFilter(RevFilter.ALL);
+            walk.markStart(head);
+
+            Map<RevCommit, List<GitVersion>> versionsByRev = versions.collect(
+                Collectors.groupingBy(GitVersion::getRevCommit));
+
+            return StreamSupport.stream(walk.spliterator(), false)
+                .flatMap(rev -> {
+                    List<GitVersion> matches = versionsByRev.get(rev);
+                    if (matches == null) {
+                        return Stream.empty();
+                    } else {
+                        // Parents can't be "nearer". Exclude them to avoid extra walking.
+                        Arrays.stream(rev.getParents()).forEach(parent -> {
+                            try {
+                                walk.markUninteresting(parent);
+                            } catch (IOException e) {
+                                throw new UncheckedIOException(e);
+                            }
+                        });
+                        return matches.stream();
+                    }
+                }).max(Comparator.comparing(GitVersion::getVersion))
+                .orElse(new GitVersion(Version.forIntegers(0, 0, 0), null));
+        } catch (IOException e) {
+            throw new UncheckedIOException(e);
+        }
+    }
+
+    private static class GitVersion {
+        private final Version version;
+        private final RevCommit revCommit;
+
+        public GitVersion(Version version, RevCommit revCommit) {
+            this.version = version;
+            this.revCommit = revCommit;
+        }
+
+        public Version getVersion() {
+            return version;
+        }
+
+        public RevCommit getRevCommit() {
+            return revCommit;
+        }
+
+        public boolean equals(Object obj) {
+            if (obj instanceof GitVersion) {
+                GitVersion other = (GitVersion) obj;
+                return Objects.equals(this.version, other.version)
+                    && Objects.equals(this.revCommit, other.revCommit);
+            } else {
+                return false;
+            }
+        }
+
+        public int hashCode() {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Add the "@Override" annotation above this method signature [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://sonarqube.ajoberstar.com/coding_rules#rule_key=squid%3AS1161)